internal/pkg/skuba/replica/replica.go

```diff
@@ -0,0 +1,406 @@
+/*
+ * Copyright (c) 2020 SUSE LLC.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package replica
+
+import (
+	"fmt"
+	"time"
+
+	appsv1 "k8s.io/api/apps/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/wait"
+	clientset "k8s.io/client-go/kubernetes"
+	"k8s.io/klog"
+	kubeadmconstants "k8s.io/kubernetes/cmd/kubeadm/app/constants"
+
+	"github.com/SUSE/skuba/internal/pkg/skuba/kubernetes"
+	"github.com/pkg/errors"
+)
+
+type affinity int
+
+const (
+	preferred affinity = iota
+	required
+)
+
+const (
+	highAvailabilitylabel = "caasp.suse.com/skuba-replica-ha"
+	minSize               = 2
+	retryCount            = 5
+	retryInterval         = kubeadmconstants.DiscoveryRetryInterval
+	retryTimeout          = kubeadmconstants.PatchNodeTimeout
+	patchAffinityRemove   = `[
+		{
+			"op":"remove",
+			"path":"/spec/template/spec/affinity"
+		}
+	]`
+	patchAffinityRequired = `{
+		"spec": {
+			"template": {
+				"spec": {
+					"affinity": {
+						"podAntiAffinity": {
+							"requiredDuringSchedulingIgnoredDuringExecution": [
+								{
+									"labelSelector": {
+										"matchExpressions": [
+											{
+												"key": "app",
+												"operator": "In",
+												"values": [
+													"%s"
+												]
+											}
+										]
+									},
+									"topologyKey": "kubernetes.io/hostname"
+								}
+							]
+						}
+					}
+				}
+			}
+		}
+	}`
+	patchAffinityPreferred = `{
+		"spec": {
+			"template": {
+				"spec": {
+					"affinity": {
+						"podAntiAffinity": {
+							"preferredDuringSchedulingIgnoredDuringExecution": [
+								{
+									"weight": 100,
+									"podAffinityTerm": {
+										"labelSelector": {
+											"matchExpressions": [
+												{
+													"key": "app",
+													"operator": "In",
+													"values": [
+														"%s"
+													]
+												}
+											]
+										},
+										"topologyKey": "kubernetes.io/hostname"
+									}
+								}
+							]
+						}
+					}
+				}
+			}
+		}
+	}`
+	patchReplicas = `{
+		"spec": {
+			"replicas": %v
+		}
+	}`
+)
+
+// Helper provide methods for replica update of deployment
+// resource that has highAvailabilitylabel.
+type Helper struct {
+	Client      clientset.Interface
+	ClusterSize int
+	Deployments *appsv1.DeploymentList
+	Deployment  *appsv1.Deployment
+	ReplicaSize int
+	MinSize     int
```

Comment:
nit: maybe we could consider not expose these variables publicly?

Reply:
sure.